devito/types/tensor.py

```diff
@@ -293,140 +216,61 @@ def values(self):
         func = vec_func(self, self)
         for j, d in enumerate(self.space_dimensions):
             comps.append(sum([getattr(self[j, i], 'd%s2' % d.name)
                               for i, d in enumerate(self.space_dimensions)]))
-        return func(name='lap_%s' % self.name, grid=self.grid,
-                    space_order=self.space_order, components=comps, time_order=to)
+        return func._new(comps)
 
     @property
     def grad(self):
         raise AttributeError("Gradient of a second order tensor not supported")
 
     def new_from_mat(self, mat):
         func = tens_func(self, self)
-        name = "%s%s" % ("_", self.name)
-        to = getattr(self, 'time_order', 0)
-        return func(name=name, grid=self.grid, space_order=self.space_order,
-                    components=mat, time_order=to, symmetric=self.is_symmetric,
-                    diagonal=self.is_diagonal)
-
+        return func._new(self.rows, self.cols, mat)
 
-class TensorTimeFunction(TensorFunction):
-    """
-    Time varying TensorFunction.
-    """
-    is_TimeDependent = True
-    is_TensorValued = True
-
-    _sub_type = TimeFunction
-    _time_position = 0
-
-    def __init_finalize__(self, *args, **kwargs):
-        super(TensorTimeFunction, self).__init_finalize__(*args, **kwargs)
-        self._time_order = kwargs.get('time_order', 1)
-
-    @classmethod
-    def __indices_setup__(cls, **kwargs):
-        return TimeFunction.__indices_setup__(grid=kwargs.get('grid'),
-                                              save=kwargs.get('save'),
-                                              dimensions=kwargs.get('dimensions'))
-
-    @property
-    def space_dimensions(self):
-        return self.indices[self._time_position+1:]
-
-    @property
-    def time_order(self):
-        return self._time_order
-
-    @property
-    def forward(self):
-        """Symbol for the time-forward state of the VectorTimeFunction."""
-        i = int(self.time_order / 2) if self.time_order >= 2 else 1
-        _t = self.indices[self._time_position]
-
-        return self.subs({_t: _t + i * _t.spacing})
-
-    @property
-    def backward(self):
-        """Symbol for the time-forward state of the VectorTimeFunction."""
-        i = int(self.time_order / 2) if self.time_order >= 2 else 1
-        _t = self.indices[self._time_position]
-
-        return self.subs({_t: _t - i * _t.spacing})
+    def classof_prod(self, other, mat):
+        try:
+            is_mat = len(mat[0]) > 1
+        except TypeError:
+            is_mat = False
+        is_time = (getattr(self, '_is_TimeDependent', False) or
+                   getattr(other, '_is_TimeDependent', False))
+        return mat_time_dict[(is_time, is_mat)]
 
 
 class VectorFunction(TensorFunction):
     """
     Vector valued space varying Function as a rank 1 tensor of Function.
     """
+
     is_VectorValued = True
     is_TensorValued = False
-
+    _is_TimeDependent = False
```

Comment:
blank line below this one

there are a few other blank-line-related unaddressed comments. I know, tedious and quite useless, but as you know I like the code to be as homogenous as possible

Reply:
fixed